libr/util/idpool.c

```diff
@@ -195,3 +200,263 @@ R_API void r_id_storage_free(RIDStorage* storage) {
+	if (storage) {
+		r_id_storage_foreach (storage->permutation, oid_free_cb, NULL);
+		r_id_storage_free (storage->permutation);
+		r_id_storage_free (storage->data);
+	}
+	free (storage);
+}
+
+R_API void *r_oids_last (ROIDStorage *storage, ut32 *odx) {
+	ut32 i, *id;
+	if (storage && storage->data && storage->data->data
+		&& storage->permutation && storage->permutation->data) {
+		for (i = 0; i < storage->permutation->top_id; i++) {
+			if (id = storage->permutation->data[i]) {
+				if (odx) {
+					*odx = i;
+				}
+				return storage->data->data[*id];
+			}
+		}
+		if (id = storage->permutation->data[i]) {
+			if (odx) {
+				*odx = i;
+			}
+			return storage->data->data[*id];
+		}
+	}
+	return NULL;
+}
+
+R_API void *r_oids_first (ROIDStorage *storage, ut32 *odx) {
+	ut32 i, *id;
+	if (storage && storage->data && storage->data->data
+		&& storage->permutation && storage->permutation->data) {
+		for (i = storage->permutation->top_id; i != 0; i--) {
+			if (id = storage->permutation->data[i]) {
+				if (odx) {
+					*odx = i;
+				}
+				return storage->data->data[*id];
+			}
+		}
+		if (id = storage->permutation->data[i]) {
+			if (odx) {
+				*odx = i;
+			}
+			return storage->data->data[*id];
+		}
+	}
+	return NULL;
+}
+
+R_API bool r_oids_foreach (ROIDStorage* storage, RIDStorageForeachCb cb, void* user) {
+	ut32 i;
+	ut32 *id;
+	if (!cb || !storage || !storage->data || !storage->data->data
+		|| !storage->permutation || !storage->permutation->data) {
+		return false;
+	}
+	for (i = storage->permutation->top_id; i != 0; i--) {
+		if ((id = storage->permutation->data[i])) {
+			if (!cb (user, storage->data->data[*id], *id)) {
+				return false;
+			}
+		}
+	}
+	if ((id = storage->permutation->data[i])) {
+		return cb (user, storage->data->data[*id], *id);
+	}
+	return true;
+}
+
+R_API bool r_oids_foreach_prev (ROIDStorage* storage, RIDStorageForeachCb cb, void* user) {
+	ut32 i;
+	ut32 *id;
+	if (!cb || !storage || !storage->data || !storage->data->data
+		|| !storage->permutation || !storage->permutation->data) {
+		return false;
+	}
+	for (i = 0; i < storage->permutation->top_id; i++) {
+		if ((id = storage->permutation->data[i])) {
+			if (!cb (user, storage->data->data[*id], *id)) {
+				return false;
+			}
+		}
+	}
+	if ((id = storage->permutation->data[i])) {
+		return cb (user, storage->data->data[*id], *id);
+	}
+	return true;
+}
+
+#if __WINDOWS__ && !__CYGWIN__
+int _cmp_wrap (void *user, const void *a, const void *b) {
+#else
+int _cmp_wrap (const void *a, const void *b, void *user) {
+#endif
+	ROIDStorage *st = (ROIDStorage *)user;
+	ut32 a_id = **((ut32 **)a);
+	ut32 b_id = **((ut32 **)b);
+	return st->cmp (r_id_storage_get (st->data, a_id),
+			r_id_storage_get (st->data, b_id));
+}
+
+R_API void r_oids_sort (ROIDStorage *storage, ROIDStorageCompareCb cmp) {
+	if (!storage || !storage->data || !storage->permutation) {
+		return;
+	}
+	storage->cmp = cmp;
+#if __WINDOWS__ && !__CYGWIN__
```

Comment:
why dat?

Reply:
qsort_r does not exist on windows, windows calls it qsort_s. qsort_s does the parameter order wrong

Reply:
apart from the name, here it seems parameters order is the same.